[4/7] Add fluent and client APIs for Evaluation Datasets

BenWilson2 opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/BenWilson2/mlflow/pull/16992?quickstart=1)

#### Install mlflow from this PR

```
# mlflow
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/16992/merge
# mlflow-skinny
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/16992/merge#subdirectory=libs/skinny
```

For Databricks, use the following command:

```
%sh curl -LsSf https://raw.githubusercontent.com/mlflow/mlflow/HEAD/dev/install-skinny.sh | sh -s pull/16992/merge
```

</p>
</details>

### Related Issues/PRs

<!-- Uncomment 'Resolve' if this PR can close the linked items. -->
<!-- Resolve --> #xxx

### What changes are proposed in this pull request?

Adds the client and fluent APIs for Evaluation Datasets.
Add a SQL-backend only error handlers for new APIs that will not work with FileStore.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

<!-- Attach code, screenshot, video used for manual testing here. -->

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

<!-- Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change. -->

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [X] No (this PR will be included in the next minor release)

BenWilson2 committed:
Evaluation Datasets Tables

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Add protos and entities for evaluation datasets

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
linting

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix MSSQL

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-tables' into eval-dataset-entity-protos

BenWilson2 committed:
fix import

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-tables' into eval-dataset-entity-protos

BenWilson2 committed:
WIP

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-tables' into eval-dataset-entity-protos

BenWilson2 committed:
Merge branch 'eval-dataset-entity-protos' into eval-dataset-sqlstore

BenWilson2 committed:
WIP

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-tables' into eval-dataset-entity-protos

BenWilson2 committed:
fix API docs formatting issue

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-entity-protos' into eval-dataset-sqlstore

BenWilson2 committed:
[3/7] Add SQL backend for Evaluation Datasets

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Add fluent and client API interfaces for Evaluation Datasets

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
lint

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

BenWilson2 committed:
fix lint issues and test failures

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
more fixes

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix docstrings

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix tests

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

BenWilson2 committed:
fix incompatible docstring compatibility test

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix for non-deterministic issue

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

BenWilson2 committed:
feedback

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-tables' into eval-dataset-entity-protos

BenWilson2 committed:
feedback

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-entity-protos' into eval-dataset-sqlstore

BenWilson2 committed:
adjustments

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
simple local import

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-entity-protos' into eval-dataset-sqlstore

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
adjustments

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

BenWilson2 committed:
test overhaul

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
debug

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
binary debug

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
debug

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
debug

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
experiment id lazy loading

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-entity-protos' into eval-dataset-sqlstore

BenWilson2 committed:
WIP

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix protos

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-entity-protos' into eval-dataset-sqlstore

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
update

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
simplify and cleanup

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

BenWilson2 committed:
WIP

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Add eval dataset tag table and related APIs

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
update schema

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

BenWilson2 committed:
WIP

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
update

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-tables' into eval-dataset-entity-protos

BenWilson2 committed:
updates

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-tables' into eval-dataset-entity-protos

BenWilson2 committed:
updates

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-entity-protos' into eval-dataset-sqlstore

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
updates

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix doc string

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-entity-protos' into eval-dataset-sqlstore

BenWilson2 committed:
Merge branch 'genai-dataset' into eval-dataset-sqlstore

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
updates

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'genai-dataset' into eval-dataset-sqlstore

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
cleanup

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
cleanup

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix schema

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

BenWilson2 committed:
fix

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fixes

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
build fix

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

harupy commented:
Can we split this PR into fluent and client?

BenWilson2 committed:
feedback

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
WIP

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

BenWilson2 committed:
wip

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
remove flush

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
try this

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix test

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

BenWilson2 committed:
fix circular imports

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
improved sql interface

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix mock

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix dependency import

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
update

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix fluent

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix backend test

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
clean it up

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'eval-dataset-sqlstore' into eval-dataset-client

B-Step62 commented on mlflow/protos/service.proto:
Can we hold these proto changes until 6th PRs? Not a big deal, but would like to review them in one place.

BenWilson2 commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -2637,6 +2654,586 @@ def _get_sql_assessment(self, session, trace_id: str, assessment_id: str) -> Sql
+        Returns:
+            List of experiment IDs associated with the dataset.
+        """
+        if not dataset_id:
```
Fixed in PR3. Will resolve once I merge in after CI validation and merge to feature branch

BenWilson2 commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -2637,6 +2654,586 @@ def _get_sql_assessment(self, session, trace_id: str, assessment_id: str) -> Sql
+                    # Extract user from record tags if provided
+                    created_by = None
+                    merged_tags = record_dict.get("tags")
+                    if merged_tags and "mlflow.user" in merged_tags:
```
Fixed in PR3

BenWilson2 commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -2637,6 +2654,586 @@ def _get_sql_assessment(self, session, trace_id: str, assessment_id: str) -> Sql
+        with self.ManagedSessionMaker() as session:
+            dataset_id = f"{self.EVALUATION_DATASET_ID_PREFIX}{uuid.uuid4().hex}"
+
+            digest = hashlib.sha256(name.encode()).hexdigest()[:8]
```
This is an excellent point. Will update!

BenWilson2 committed:
Merge branch 'genai-dataset' into eval-dataset-client

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 commented on mlflow/genai/datasets/__init__.py:
```diff
@@ -6,20 +6,222 @@
+
+        delete_evaluation_dataset(name="catalog.schema.table")
+    """
+    if is_in_databricks_runtime():
```
good call!

BenWilson2 committed:
feedback and enhancements

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
attempt fix

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
attempt

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Merge branch 'genai-dataset' into eval-dataset-client

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
API unification

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
protos

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
attempt fix for mssql docker

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix attempt

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fixes

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
eval-dataset-client

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

TomeHirata commented on mlflow/store/tracking/sqlalchemy_store.py:
```diff
@@ -3015,16 +3014,16 @@ def _compute_evaluation_dataset_digest(self, name: str, last_update_time: int) -
-        tags: Optional[dict[str, str]] = None,
-        experiment_ids: Optional[list[str]] = None,
+        tags: dict[str, str] | None = None,
+        experiment_id: list[str] | None = None,
```
Shouldn't we keep this argument as `experiment_ids` as this is a list of experiment id?

TomeHirata commented on mlflow/genai/datasets/__init__.py:
```diff
@@ -6,60 +6,357 @@
+    else:
+        from mlflow.tracking.client import MlflowClient
+
+        client = MlflowClient()
+        return client.create_dataset(
+            name=name,
+            experiment_id=experiment_ids,
+            tags=tags,
+        )
```
nit:
```suggestion
        return MlflowClient().create_dataset(
            name=name,
            experiment_id=experiment_ids,
            tags=tags,
        )
```

TomeHirata commented on mlflow/genai/datasets/__init__.py:
```diff
@@ -6,60 +6,357 @@
+
+        from mlflow.tracking.client import MlflowClient
+
+        client = MlflowClient()
+        client.delete_dataset(dataset_id)
```
nit:
```suggestion
        MlflowClient().delete_dataset(dataset_id)
```

TomeHirata commented on mlflow/genai/datasets/__init__.py:
```diff
@@ -6,60 +6,357 @@
+
+    from mlflow.tracking.client import MlflowClient
+
+    client = MlflowClient()
+    client.set_dataset_tags(dataset_id, tags)
```
nit:
```suggestion
    MlflowClient().set_dataset_tags(dataset_id, tags)
```

TomeHirata commented on mlflow/genai/utils/trace_utils.py:
```diff
@@ -80,6 +79,16 @@ def __exit__(self, exc_type, exc_value, traceback):
         NoOpTracer.start_span = self.original
 
 
+def is_none_or_nan(value: Any) -> bool:
```
q: why can't we reuse the one in `mlflow.genai.evaluation.utils`?

BenWilson2 replied:
This creates a circular import that is definitely fixable, but involves some significant refactoring that I feel will be fragile to maintain and will create headaches for the addition of further APIs in the genai namespace. Porting this util function over, while not DRY, is definitely the lesser of two evils in my opinion.

TomeHirata commented on mlflow/utils/annotations.py:
```diff
@@ -226,3 +226,26 @@ def wrapper(*args, **kwargs):
+    )
+    func.__doc__ = notice + func.__doc__ if func.__doc__ else notice
+
+    func._requires_sql_backend = True
```
q: will we implement the validation in the following PRs?

TomeHirata commented on mlflow/utils/annotations.py:
```diff
@@ -226,3 +226,26 @@ def wrapper(*args, **kwargs):
     return wrapper
+
+
+def requires_sql_backend(func):
```
This might be useful for webhook cc: @harupy

B-Step62 reviewed (approved):
The comment about the param validation and client API is not blocker. Feel free to ignore if it takes more than 15 mins to adopt.

BenWilson2 committed:
updates and optimizations

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
fix test lazy import issue

Signed-off-by: Ben Wilson <benjamin.wilson@databricks.com>

BenWilson2 committed:
Fix mock.patch import issue in test_evaluate_custom_metric_incorrect_return_formats

The test was failing in CI with AttributeError because mock.patch() with the string path 'mlflow.models.evaluation.utils.metric._logger.warning' doesn't work - the 'utils' module is not accessible as an attribute of mlflow.models.evaluation.

Fixed by importing the metric module directly and using mock.patch.object() to patch the logger attribute directly on the module object.

BenWilson2 commented:
The test failures here are unrelated to any changes and all outstanding comments have been addressed (with the exception of removal of client API interfaces, which I feel we should keep for consistency with other similar API interfaces within MLflow for top-level entities).
